TinyForward/TinyForward.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 10/22/11               \" DATE 
.Dt TinyForward 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm TinyForward,
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.Nm Other_name_for_same_program(),
.Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd This line parsed for whatis database.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl abcd              \" [-abcd]
.Op Fl a Ar path         \" [-a path] 
.Op Ar file              \" [file]
.Op Ar                   \" [file ...]
.Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
Use the .Nm macro to refer to your program throughout the man page like such:
.Nm
Underlining is accomplished with the .Ar macro like this:
.Ar underlined text .
.Pp                      \" Inserts a space
A list of items with descriptions:
.Bl -tag -width -indent  \" Begins a tagged list 
.It item a               \" Each item preceded by .It macro
Description of item a
.It item b
Description of item b
.El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl a                 \"-a flag as a list item
Description of -a flag
.It Fl b
Description of -b flag
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/Users/joeuser/Library/really_long_file_name" -compact
.It Pa /usr/share/file_name
FILE_1 description
.It Pa /Users/joeuser/Library/really_long_file_name
FILE_2 description
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr a 1 , 
.Xr b 1 ,
.Xr c 1 ,
.Xr a 2 ,
.Xr b 2 ,
.Xr a 3 ,
.Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner

TinyForward/tinyforward.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <arpa/inet.h>

#define HOST    "0.0.0.0"
#define PORT    5560
#define BUFFER_SIZE  10

int connectedClients, connectedServers;

typedef enum {
    ClientSocket,
    ServerSocket,
    ListenerSocket
} SocketType;

typedef struct ProxySocket Socket;

struct ProxySocket {
    Socket *prevousSocket;
    SocketType type;
    int id;
    struct ProxySocket *destinationSocket;
    char *bytesRead;
    char *bytesWritten;
    ssize_t readSize;
    ssize_t writeSize;
    Socket *nextSocket;
};

Socket *lastSocket;
fd_set readSet, tempReadSet, writeSet;

Socket *addSocket(int socketId, SocketType type)
{
    static int open = 0;
    fprintf(stderr, "Open sockets: %d", ++open);
    Socket *newSocket;
    newSocket = (Socket*)malloc(sizeof(Socket));
    newSocket->destinationSocket = NULL;
    newSocket->bytesRead = NULL;
    newSocket->bytesWritten = NULL;
    newSocket->readSize = 0;
    newSocket->writeSize = 0;
    newSocket->nextSocket = NULL;
    newSocket->type = type;
    newSocket->id = socketId;
    
    //Add it to the linked list
    newSocket->prevousSocket = lastSocket;
    lastSocket->nextSocket = newSocket;
    lastSocket = newSocket;
    
    return newSocket;
}

void removeSocket(Socket *socket)
{
    static int close = 0;
    fprintf(stderr, "Closed sockets: %d", ++close);
    // remove from linked list
    socket->prevousSocket->nextSocket = socket->nextSocket;
    if(socket->nextSocket != NULL)
        socket->nextSocket->prevousSocket = socket->prevousSocket;
    if(lastSocket == socket)
        lastSocket = socket->prevousSocket;
    free(socket->bytesRead);
    free(socket);
}

int getHostAddr (struct sockaddr_in *name, const char *hostName, uint16_t port)
{
    struct hostent *hostinfo;
    
    name->sin_family = AF_INET;
    name->sin_port = htons (port);
    hostinfo = gethostbyname (hostName);
    if (hostinfo == NULL)
    {
        return -1;
    }
    name->sin_addr = *(struct in_addr *) hostinfo->h_addr;
    return 0;
}

int
read_from_client (Socket *client)
{
    char *buffer = (char*)malloc(BUFFER_SIZE);
    ssize_t nbytes;
    
    nbytes = recv(client->id, buffer, BUFFER_SIZE, 0);
    fprintf(stderr, "Bytes: %ld", nbytes);
    if (nbytes < 0)
    {
        free(buffer);
        return -1;
        /* Read error. */
        //perror ("read");
        //exit (EXIT_FAILURE);
    }
    else if (nbytes == 0)
    {
        free(buffer);
        return 0;
    }
    else
    {
        /* Data read. */
        client->bytesRead = realloc(client->bytesRead, (client->bytesRead ? strlen(client->bytesRead) : 0) + strlen(buffer) + sizeof(char));
        strncat(client->bytesRead, buffer, nbytes);
        if(nbytes < BUFFER_SIZE){
            fprintf (stderr, "Server: whole message: '%s'\n", client->bytesRead);
        }
        //fprintf (stderr, "Server: got message: '%s'\n", buffer);
        free(buffer);
        return 0;
    }
}

int getHostFromHeaders(char *headers, char **hostName, int *port)
{
    static size_t prefixLength = strlen("Host: ");
    const char *hostHeader;
    size_t length;
    char *portStr;
    
    hostHeader = strstr(headers, "Host: ");
    if(hostHeader == NULL)
        return -1;
    length = strcspn(hostHeader, "\r\n");
    *hostName = (char *)malloc(length - prefixLength + 1); // room for null terminator
    **hostName = '\0'; // so strncat will start at beginning
    strncat(*hostName, hostHeader + prefixLength, length - prefixLength);
    
    *port = 80;
    if(strchr(*hostName, ':') != NULL)
    {
        strtok (*hostName, ":"); // cut string at colon
        portStr = strtok (NULL, ":"); // get port number
        *port = atoi (portStr); // convert ASCII sting to number
    }
    
    return 0;
}

int connectToServer(Socket *clientSocket)
{
    struct sockaddr_in name;
    char *hostName;
    int port;
    int server;
    Socket *serverSocket;
    
    if(getHostFromHeaders(clientSocket->bytesRead, &hostName, &port) < 0)
    {
        fprintf(stderr, "Cannot get hostname from headers.");
        return -1;
    }
    
    server = socket (PF_INET, SOCK_STREAM, 0);
    if(server < 0)
    {
        fprintf(stderr, "Cannot create client socket.");
        return -1;
    }
    if(getHostAddr(&name, hostName, port) < 0)
    {
        fprintf(stderr, "Cannot get host address from host name: %s and port: %d.", hostName, port);
        return -1;
    }
    if (connect (server, (struct sockaddr *) &name, sizeof (name)) < 0)
    {
        fprintf(stderr, "Cannot connect to server: %s on port: %d.", hostName, port);
        return -1;
    }
    
    serverSocket = addSocket(server, ServerSocket);
    
    //Associate with client
    clientSocket->destinationSocket = serverSocket;
    serverSocket->destinationSocket = clientSocket;
    
    //Allow socket to be read
    FD_SET (serverSocket->id, &readSet);
    
    // Stop reading client socket
    //FD_CLR (clientSocket->id, &readSet);
    
    free(hostName);
    
    // Socket is ready to be written to
    //clientSocket->readSize = strlen(clientSocket->bytesRead); // This allows checking to see if socket is ready
    //FD_SET(clientSocket->destinationSocket->id, &writeSet);
    return 0;
}

int readSocket (Socket *socket)
{
    // parse input
    // if done with parseing
    // set bytesRead
    // if no connection, connect to server
    // add server to read set
    // else continue
    char *buffer = (char*)malloc(BUFFER_SIZE);
    ssize_t count;
    
    count = recv(socket->id, buffer, BUFFER_SIZE, 0);
    if (count < 0) // error
    {
        // set server socket ready to write
        // if bytesRead != null
        free(buffer);
        return -1;
    }
    else if (count == 0) // read no bytes
    {
        free(buffer);
        //if(client->readSize == 0 && client->bytesRead != NULL){
        //    return readyForServer(client);
        //}
        return -1;
    }
    else
    {
        // if bytes are written, we can release old memory and malloc new memory
        // if not yet written, add on to the memory block
        // resize the bytesRead array, but if it's null just alloc. Don't forget to make room for the trailing null byte
        if(socket->writeSize >= socket->readSize)
        {
            free(socket->bytesRead);
            socket->bytesRead = NULL;
            socket->readSize = 0;
            socket->bytesWritten = NULL;
            socket->writeSize = 0;
        }
        socket->bytesRead = realloc(socket->bytesRead, (socket->bytesRead ? strlen(socket->bytesRead) : 0) + strlen(buffer) + sizeof(char));
        if(socket->readSize == 0)
            *socket->bytesRead = '\0'; // Reset the buffer
        socket->readSize += count;
        strncat(socket->bytesRead, buffer, count); // use count to make sure garbage data isn't copied
        free(buffer);
        if(socket->type == ClientSocket)
        {
            if(count < BUFFER_SIZE)
            {
                // set server socket ready to write
                if(socket->destinationSocket == NULL)
                {
                    if(connectToServer(socket) < 0) // todo fix case when count = buffer_size
                    {
                        return -1;
                    }
                }
                
                FD_SET(socket->destinationSocket->id, &writeSet); // we can write to it
            }
            
            return 0;
        }
        // server sockets
        if(socket->destinationSocket != NULL)
        {
            FD_SET(socket->destinationSocket->id, &writeSet); // we can write to it
        }
        //fprintf (stderr, "Server: got message: '%s'\n", buffer);
    }
    
    return 0;
}

/*
int readServer (Socket *server)
{
    // read data to buffer
    // associate buffer with client
    // add client to write set
    // if end, remove self from read set
    if(server->destinationSocket == NULL)
        return -1;
    
    if(server->bytesRead == NULL)
        server->bytesRead = malloc(BUFFER_SIZE);
    
    server->readSize = recv(server->id, server->bytesRead, BUFFER_SIZE, 0);
    if (server->readSize <= 0)
    {
        // set server socket ready to write
        // if bytesRead != null
        return -1;
    }
    else
    {
        FD_SET(server->destinationSocket->id, &writeSet);
        FD_CLR(server->id, &readSet);
    }
    
    return 0;
}
*/

int writeSocket (Socket *socket)
{
    // write buffer to socket
    // remove self from write set
    size_t count;
    if(socket->destinationSocket == NULL)
    {
        return -1;
    }
    else if(socket->destinationSocket->bytesRead == NULL)
    {
        return 0;
    }
    else if(socket->writeSize >= socket->readSize)
    {
        return -1;
    }
    if(socket->destinationSocket->bytesWritten == 0)
    {
        socket->destinationSocket->bytesWritten = socket->destinationSocket->bytesRead;
    }
    count = send(socket->id, socket->destinationSocket->bytesWritten, socket->destinationSocket->readSize - socket->destinationSocket->writeSize, 0);
    socket->destinationSocket->bytesWritten += count;
    socket->destinationSocket->writeSize += count;
    
    //socket->destinationSocket->readSize = 0;
    
    if(socket->destinationSocket->writeSize >= socket->destinationSocket->readSize) // we read everything we can
        FD_CLR(socket->id, &writeSet);
    else
        FD_SET(socket->id, &writeSet); // TODO: Find out if this can be removed
    //FD_SET (socket->destinationSocket->id, &readSet);
    return 0;
}

int createListenerSocket (const char *host, uint16_t port)
{
    int sock;
    struct sockaddr_in name;
    
    if ((sock = socket (PF_INET, SOCK_STREAM, 0)) < 0) // make the socket
    {
        perror ("create");
        exit (EXIT_FAILURE);
    }
    
    // define the socket parameters
    name.sin_family = AF_INET;
    name.sin_port = htons (port);
    if(inet_pton (AF_INET, host, &name.sin_addr.s_addr) < 1) // convert host string to IP type
    {
        perror ("parse");
        close (sock);
        exit (EXIT_FAILURE);
    }
    if (bind (sock, (struct sockaddr *) &name, sizeof (name)) < 0) // bind the socket
    {
        perror ("bind");
        close (sock);
        exit (EXIT_FAILURE);
    }
    
    return sock;
}

int main (void)
{
    Socket *listenerSocket;
    Socket *currentSocket;
    //int socket;
    int newClient;
    //struct sockaddr_in client;
    //socklen_t size;
    
    // create listener socket
    listenerSocket = (Socket*)malloc(sizeof(Socket));
    listenerSocket->destinationSocket = NULL;
    listenerSocket->bytesRead = NULL;
    listenerSocket->nextSocket = NULL;
    listenerSocket->prevousSocket = NULL;
    listenerSocket->type = ListenerSocket;
    listenerSocket->id = createListenerSocket (HOST, PORT);
    
    if (listen (listenerSocket->id, 1) < 0) // start listening
    {
        perror ("listen");
        close (listenerSocket->id);
        exit (EXIT_FAILURE);
    }
    
    lastSocket = listenerSocket; // the last socket is the first one
    
    // initialize the sets
    FD_ZERO (&readSet);
    //FD_ZERO (&tempReadSet);
    FD_ZERO (&writeSet);
    FD_SET (listenerSocket->id, &readSet);

    
    do
    {
        tempReadSet = readSet; // copy the set because select() destroys our input
        if (select (FD_SETSIZE, &tempReadSet, &writeSet, NULL, NULL) < 0)
        {
            perror ("select");
            close (listenerSocket->id);
            exit (EXIT_FAILURE);
        }
        
        /* Service all the sockets with input pending. */
        assert(listenerSocket != NULL);
        for (currentSocket = listenerSocket; currentSocket != NULL; currentSocket = currentSocket->nextSocket)
        {
            if (FD_ISSET (currentSocket->id, &tempReadSet)) // the socket can be read
            {
                FD_CLR(currentSocket->id, &tempReadSet);
                if (currentSocket->type == ListenerSocket)
                {
                    // get connection request
                    //size = sizeof (client);
                    newClient = accept (listenerSocket->id, NULL, NULL);
                                  //(struct sockaddr *) &client,
                                  ///&size);
                    if (newClient < 0)
                    {
                        fprintf(stderr, "Error: %d", errno);
                        perror ("accept");
                        exit (EXIT_FAILURE);
                    }
                    
                    addSocket(newClient, ClientSocket);
                    
                    fprintf (stderr, "New connection.");
                    FD_SET (newClient, &readSet);
                    break;
                    // TODO: bug! If this break isn't here, program will infinite loop when too many clients try to connect at once
                }
                else
                {
                    if (readSocket (currentSocket) < 0)
                    {
                        if(currentSocket->writeSize >= currentSocket->readSize) // make sure everything's written
                        {
                            fprintf (stderr, "Closing socket.");
                            close (currentSocket->id); // close connection
                            FD_CLR (currentSocket->id, &readSet);
                            FD_CLR (currentSocket->id, &writeSet);
                            if(currentSocket->destinationSocket != NULL)
                            {
                                FD_CLR (currentSocket->destinationSocket->id, &writeSet);
                                currentSocket->destinationSocket->destinationSocket = NULL;
                            }
                            removeSocket(currentSocket);
                            currentSocket = NULL;
                            break; // it's easier to lose some speed than try to free the pointer elsewhere
                        }
                    }
                }
            }
            if (FD_ISSET (currentSocket->id, &writeSet)) // the socket can be written
            {
                if (writeSocket (currentSocket) < 0)
                {
                    
                }
            }
            assert(currentSocket->nextSocket != currentSocket);
        }
    }while(1);
    
    free(listenerSocket);
}